File: 7. 10001st Prime/solution.py
from math import sqrt, log

def n_prime(n):
    upper_bound = int(n * (log(n) + log(log(n))))

    sieve = [False, False] + [True] * (upper_bound - 1)

    for i in range(2, int(sqrt(upper_bound)) + 1):
        if sieve[i]:
            for j in range(i * i, upper_bound + 1, i):
                sieve[j] = False

    count = 0
    for num, is_prime in enumerate(sieve):
        if is_prime:
            count += 1
            if count == n:
                return num

print(n_prime(10001))

File: 7. 10001st Prime/solution.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(nPrime(10001))
}

func nPrime(n int) int {
	upperBound := int(math.Ceil(float64(n) * (math.Log(float64(n)) + math.Log(math.Log(float64(n))))))

	sieve := make([]bool, upperBound+1)
	for i := range sieve {
		sieve[i] = true
	}
	sieve[0], sieve[1] = false, false

	for i := 2; i <= upperBound; i++ {
		if sieve[i] {
			for j := i * i; j <= upperBound; j += i {
				sieve[j] = false
			}
		}
	}

	count := 0

	for num := 0; num < len(sieve); num++ {
		if sieve[num] {
			count++
			if count == n {
				return num
			}
		}
	}

	return -1
}
